engine/src/main/java/org/terasology/rendering/nui/layers/mainMenu/selectModulesScreen/SelectModulesScreen.java

```diff
@@ -359,50 +362,141 @@ public String get() {
+            localOnlyCheckbox.subscribe(e -> {
+                localOnlyCheckbox.setChecked(!selectModulesConfig.isLocalOnlyChecked());
+                selectModulesConfig.setIsLocalOnlyChecked(!selectModulesConfig.isLocalOnlyChecked());
+                filterModules();
+            });
+        }
         WidgetUtil.trySubscribe(this, "close", button -> triggerBackAnimation());
     }
 
-    private void prepareModuleList(boolean checked) {
-        if (selectModulesConfig.isChecked()) {
-            Iterator<ModuleSelectionInfo> iter = sortedModules.iterator();
-            while (iter.hasNext()) {
-                if (!iter.next().isPresent()) {
-                    iter.remove();
-                }
+     private void filterModules() {
+        sortedModules.clear();
+        sortedModules.addAll(allSortedModules);
+        if(selectModulesConfig.isAssetplayChecked() || selectModulesConfig.isAugmentationChecked() || selectModulesConfig.isGameplayChecked() || selectModulesConfig.isSpecialChecked() || selectModulesConfig.isWorldChecked() || selectModulesConfig.isLibraryChecked()) {
+            advancedModuleFilter();
+        }
+         if (selectModulesConfig.isLocalOnlyChecked()) {
+            localModuleFilter();
+        }
+        filterText();
+    }
+
+    private void localModuleFilter() {
+        Iterator<ModuleSelectionInfo> iter = sortedModules.iterator();
+        while (iter.hasNext()) {
+            if (!iter.next().isPresent()) {
+                iter.remove();
             }
         }
     }
 
-    private void filterText(String newText) {
-        sortedModules.clear();
-        for (ModuleSelectionInfo m : allSortedModules) {
-            if (m.getMetadata().getDisplayName().toString().toLowerCase().contains(newText.toLowerCase())) {
-                sortedModules.add(m);
+    private void filterText() {
+        String newText = moduleSearch.getText();
+        Iterator<ModuleSelectionInfo> iter = sortedModules.iterator();
+        while (iter.hasNext()) {
+            if (!iter.next().getMetadata().getDisplayName().toString().toLowerCase().contains(newText.toLowerCase())) {
+                iter.remove();
             }
         }
     }
 
+
+    private void advancedModuleFilter() {
+        Iterator<ModuleSelectionInfo> iter = sortedModules.iterator();
+        while (iter.hasNext()) {
+            ModuleSelectionInfo m = iter.next();
+            Module module;
+            if(m.isPresent()) {
+                module = moduleManager.getRegistry().getLatestModuleVersion(m.getMetadata().getId());//(m.getOnlineVersion() == null) ? m.getLatestVersion() : m.getOnlineVersion();
+            }
+            else {
+                module = (m.getOnlineVersion() == null) ? m.getLatestVersion() : m.getOnlineVersion();
+            }
+
+            if(selectModulesConfig.isLibraryChecked() && StandardModuleExtension.isLibraryModule(module)) {
```

Comment:
I think it would be better to get the different types of modules in a `Set` so you can do simply call `removeAll()` or `retainAll()` instead of manually iterating over all possible values. This is also somewhat error prone, in particular when a new type is added.